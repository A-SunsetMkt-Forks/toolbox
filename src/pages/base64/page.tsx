import { PageContainer } from '@ant-design/pro-components';
import { Box } from '@mui/system';
import type { SegmentedProps } from 'antd';
import { Button, Input, Segmented, Space } from 'antd';
import { useCallback, useState } from 'react';

import TextResultBox from '@/components/TextResultBox.tsx';
import type { BaseType } from '@/lib/BaseEncoder.ts';
import { decode as baseDecode, encode as baseEncode } from '@/lib/BaseEncoder.ts';

export default function Page() {
  const [type, setType] = useState<BaseType>('base-64');
  const [plain, setPlain] = useState<string>('');
  const [target, setTarget] = useState<string>('');

  const segmentedOptions: SegmentedProps['options'] = [
    { label: 'Base16', value: 'base-16' },
    { label: 'Base32', value: 'base-32', disabled: false },
    { label: 'Z Base32', value: 'z-base-32', disabled: false },
    { label: 'Base36', value: 'base-36', disabled: false },
    { label: 'Base58', value: 'base-58', disabled: false },
    { label: 'Base58 Check', value: 'base-58-check', disabled: false },
    { label: 'Base64', value: 'base-64' },
  ];

  const encode = useCallback(() => {
    const result = baseEncode(type, plain);
    setTarget(result);
  }, [type, plain]);

  const decode = useCallback(() => {
    const result = baseDecode(type, plain);
    setTarget(result);
  }, [type, plain]);

  return (
    <PageContainer token={{ paddingInlinePageContainerContent: 10, paddingBlockPageContainerContent: 10 }}>
      <Box sx={{ mt: 2, mb: 2 }}>
        <Segmented
          block={true}
          value={type}
          options={segmentedOptions}
          onChange={(value) => {
            setType(value as BaseType);
          }}
        />
      </Box>
      <Box sx={{ mt: 2 }}>
        <Input.TextArea
          value={plain}
          onChange={(event) => {
            setPlain(event.target.value);
          }}
          placeholder="文本"
          autoSize={{ minRows: 5, maxRows: 10 }}
          showCount
        />
      </Box>
      <Box sx={{ mt: 2 }}>
        <Space.Compact>
          <Button onClick={encode}>编码</Button>
          <Button onClick={decode}>解码</Button>
        </Space.Compact>
      </Box>
      <div className="mt-2">
        <TextResultBox title="结果" text={target} />
      </div>
    </PageContainer>
  );
}
