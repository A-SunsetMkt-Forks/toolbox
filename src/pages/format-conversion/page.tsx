import { PageContainer } from '@ant-design/pro-components';
import type { SelectProps } from 'antd';
import { Button, Card, Col, Row, Select } from 'antd';
import XML from 'fast-xml-parser';
import * as LosslessJSON from 'lossless-json';
import { useState } from 'react';
import AceEditor from 'react-ace';
import type { AceOptions } from 'react-ace/types';
import YAML from 'yaml';

import Iconify from '@/components/Iconify';
import { format } from '@/utils/formatter.ts';

import 'ace-builds/src-noconflict/ext-searchbox';
import 'ace-builds/src-noconflict/mode-json';
import 'ace-builds/src-noconflict/mode-xml';
import 'ace-builds/src-noconflict/mode-yaml';
import 'ace-builds/src-noconflict/theme-monokai';

const CodeTypeOptions: SelectProps['options'] = [
  { label: 'JSON', value: 'json' },
  { label: 'YAML', value: 'yaml' },
  { label: 'XML', value: 'xml' },
];

const options: AceOptions = {
  useWorker: false,
  tabSize: 2,
  wrap: true,
  showLineNumbers: true,
};

const targetAceEditorOptions: AceOptions = {
  useWorker: false,
  tabSize: 2,
  wrap: true,
  showLineNumbers: true,
};
const FormatConversionPage = () => {
  const [sourceType, setSourceType] = useState<string>('json');
  const [targetType, setTargetType] = useState<string>('yaml');
  const [sourceValue, setSourceValue] = useState<string>('');
  const [targetValue, setTargetValue] = useState<string>('');

  const onClickSyncAltBtn = () => {
    const targetTypeTmp = targetType;
    setTargetType(sourceType);
    setSourceType(targetTypeTmp);
  };
  const onClickConversionBtn = () => {
    if (sourceValue === undefined || sourceValue === '') {
      return;
    }

    let sourceObj = null;

    // 解析 source
    try {
      if (sourceType === 'json') {
        sourceObj = LosslessJSON.parse(sourceValue, null, LosslessJSON.parseNumberAndBigInt);
      } else if (sourceType === 'yaml') {
        sourceObj = YAML.parse(sourceValue, { intAsBigInt: true });
      } else if (sourceType === 'xml') {
        const xmlParser = new XML.XMLParser();
        sourceObj = xmlParser.parse(sourceValue);
      }
    } catch (e) {
      alert(e);
    }

    // 设置 target
    try {
      if (targetType === 'yaml') {
        format(YAML.stringify(sourceObj), 'yaml').then((val) => {
          setTargetValue(val);
        });
      } else if (targetType === 'json') {
        format(LosslessJSON.stringify(sourceObj) || '', 'json').then((val) => {
          setTargetValue(val);
        });
      } else if (targetType === 'xml') {
        const xmlBuilder = new XML.XMLBuilder({
          ignoreAttributes: false,
          format: true,
          oneListGroup: true,
          suppressEmptyNode: true,
        });
        setTargetValue(xmlBuilder.build(sourceObj));
      }
    } catch (e) {
      alert(e);
    }
  };
  return (
    <PageContainer title={false} className="bg-gray-50">
      <div className="mx-auto h-[calc(100dvh-56px)] max-w-screen-xl p-4 md:h-screen">
        <Card>
          <Row gutter={{ xs: 8, sm: 8, md: 16 }}>
            <Col span={8} sm={6} md={4}>
              <Select options={CodeTypeOptions} value={sourceType} onSelect={setSourceType} className="w-full" />
            </Col>
            <Col>
              <Button icon={<Iconify icon="material-symbols:sync-alt" />} onClick={onClickSyncAltBtn} />
            </Col>
            <Col span={8} sm={6} md={4}>
              <Select options={CodeTypeOptions} value={targetType} onSelect={setTargetType} className="w-full" />
            </Col>
            <Col span={4} sm={4} md={2}>
              <Button onClick={onClickConversionBtn}>转换</Button>
            </Col>
          </Row>
        </Card>
        <Row className="mt-4" gutter={[16, 16]}>
          <Col span={24} md={12}>
            <AceEditor
              mode={sourceType}
              theme="monokai"
              width="100%"
              height=""
              className="h-[calc(50vh-100px)] md:h-[calc(100vh-130px)]"
              setOptions={options}
              value={sourceValue}
              onChange={setSourceValue}
            />
          </Col>
          <Col span={24} md={12}>
            <AceEditor
              mode={targetType}
              theme="monokai"
              width="100%"
              height=""
              className="h-[calc(50dvh-100px)] md:h-[calc(100dvh-130px)]"
              setOptions={targetAceEditorOptions}
              value={targetValue}
              onChange={setTargetValue}
            />
          </Col>
        </Row>
      </div>
    </PageContainer>
  );
};

export default FormatConversionPage;
