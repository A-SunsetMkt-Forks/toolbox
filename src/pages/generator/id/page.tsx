import { PageContainer } from '@ant-design/pro-components';
import { createId } from '@paralleldrive/cuid2';
import { TwitterSnowflake } from '@sapphire/snowflake';
import { useCounter } from 'ahooks';
import { Button, Card, Segmented } from 'antd';
import type { SegmentedLabeledOption } from 'antd/es/segmented';
import ObjectID from 'bson-objectid';
import { nanoid } from 'nanoid';
import type { SegmentedRawOption, SegmentedValue } from 'rc-segmented';
import type { FC } from 'react';
import { useEffect, useState } from 'react';
import { ulid } from 'ulid';
import { v4 as uuidv4 } from 'uuid';

import TextResultBox from '../../../components/TextResultBox.tsx';

const UidPage: FC = () => {
  const [type, setType] = useState<SegmentedValue>('cuid');
  const [result, setResult] = useState<string>('');
  const [counter, { inc }] = useCounter(0);

  const options: (SegmentedRawOption | SegmentedLabeledOption)[] = [
    { label: 'Cuid2', value: 'cuid' },
    { label: 'BSON ObjectID', value: 'objectid' },
    { label: 'Nano ID', value: 'nanoid' },
    { label: 'ULID', value: 'ulid' },
    { label: 'UUID', value: 'uuid' },
    { label: 'Snowflake ID', value: 'snowflakeid' },
  ];

  useEffect(() => {
    if (type === 'cuid') {
      setResult(createId);
    } else if (type === 'objectid') {
      setResult(ObjectID().toHexString());
    } else if (type === 'nanoid') {
      setResult(nanoid());
    } else if (type === 'ulid') {
      setResult(ulid());
    } else if (type === 'uuid') {
      setResult(uuidv4());
    } else if (type === 'snowflakeid') {
      setResult(TwitterSnowflake.generate().toString());
    }
  }, [type, counter]);
  return (
    <PageContainer title={false} className="mx-auto max-w-screen-lg pt-4">
      <Card>
        <Segmented block={true} options={options} value={type} onChange={(value) => setType(value)} />
        <div className="mt-4">
          <Button onClick={() => inc()}>刷新</Button>
        </div>
        <div className="mt-4">
          <TextResultBox title="结果" text={result} />
        </div>
      </Card>
    </PageContainer>
  );
};
export default UidPage;
