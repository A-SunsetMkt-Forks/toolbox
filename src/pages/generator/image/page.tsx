import { PageContainer, ProForm, ProFormDigit, ProFormSelect, ProFormText } from '@ant-design/pro-components';
import { Box } from '@mui/system';
import type { TabsProps } from 'antd';
import { Col, ColorPicker, Form, Row, Tabs } from 'antd';
import type { FC } from 'react';
import { useEffect, useState } from 'react';

import type { PlaceholderConfig } from '@/components/placeholder';
import DummyImage from '@/components/placeholder/DummyImage.tsx';
import HrefLuImage from '@/components/placeholder/HrefLuImage.tsx';
import PlaceholderImage from '@/components/placeholder/PlaceholderImage.tsx';

const defaultValues = {
  width: 320,
  height: 320,
  suffix: '.png',
  bgColor: '#cccccc',
  textColor: '#969696',
  text: 'img',
};

const suffixOptions = [
  { label: '.png', value: '.png' },
  { label: '.jpg', value: '.jpg' },
  { label: '.gif', value: '.gif' },
];

const PlaceholderImg: FC = () => {
  const [form] = Form.useForm();
  const [placeholderConfig, setPlaceholderConfig] = useState<PlaceholderConfig>(defaultValues);

  const bgColor = Form.useWatch('bgColor', { form, preserve: true });
  const textColor = Form.useWatch('textColor', { form, preserve: true });

  useEffect(() => {
    const values = form.getFieldsValue();
    setPlaceholderConfig(values);
  }, [bgColor, form, textColor]);

  const tabsProps: TabsProps = {
    defaultActiveKey: 'HrefLuImage',
    items: [
      {
        key: 'HrefLuImage',
        label: 'iph.href.lu',
        children: <HrefLuImage {...placeholderConfig} />,
      },
      {
        key: 'PlaceholderImage',
        label: 'via.placeholder.com',
        children: <PlaceholderImage {...placeholderConfig} />,
      },
      {
        key: 'DummyImage',
        label: 'dummyimage.com',
        children: <DummyImage {...placeholderConfig} />,
      },
    ],
  };
  return (
    <PageContainer>
      <ProForm
        submitter={false}
        form={form}
        initialValues={defaultValues}
        onValuesChange={(_, values) => {
          setPlaceholderConfig(values);
        }}
      >
        <Row gutter={16}>
          <Col xs={24} sm={24} md={12} lg={8}>
            <ProFormDigit label="宽度" name="width" />
          </Col>
          <Col xs={24} sm={24} md={12} lg={8}>
            <ProFormDigit label="高度" name="height" />
          </Col>
          <Col xs={24} sm={24} md={12} lg={8}>
            <ProFormSelect label="后缀名" name="suffix" options={suffixOptions} allowClear={false} />
          </Col>
          <Col xs={24} sm={24} md={12} lg={8}>
            <ProFormText
              label="背景颜色"
              name="bgColor"
              fieldProps={{
                suffix: (
                  <ColorPicker
                    value={bgColor}
                    onChange={(_, hex) => {
                      form.setFieldValue('bgColor', hex);
                    }}
                  />
                ),
              }}
              allowClear={false}
            />
          </Col>
          <Col xs={24} sm={24} md={12} lg={8}>
            <ProFormText
              label="文本颜色"
              name="textColor"
              fieldProps={{
                suffix: (
                  <ColorPicker
                    value={textColor}
                    onChange={(_, hex) => {
                      form.setFieldValue('textColor', hex);
                    }}
                  />
                ),
              }}
              allowClear={false}
            />
          </Col>
          <Col xs={24} sm={24} md={12} lg={8}>
            <ProFormText label="文本" name="text" allowClear={false} fieldProps={{ style: { height: '42px' } }} />
          </Col>
        </Row>
      </ProForm>
      <Box>
        <Tabs {...tabsProps} />
      </Box>
    </PageContainer>
  );
};

export default PlaceholderImg;
