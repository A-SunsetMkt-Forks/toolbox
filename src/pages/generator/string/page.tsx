import { PageContainer } from '@ant-design/pro-components';
import { Button, Card, Checkbox, Form, Input, Slider } from 'antd';
import { useEffect, useState } from 'react';

import Iconify from '@/components/Iconify';
import InputCopyable from '@/components/InputCopyable.tsx';
import type { GeneratorOptions } from '@/scripts/RandomStringUtils.ts';
import { generator } from '@/scripts/RandomStringUtils.ts';

const initialOptions: GeneratorOptions = {
  numeric: true,
  lowercase: true,
  uppercase: true,
  symbol: false,
  symbolChars: '!@#$%^&*',
};
export const Component = () => {
  const [randomString, setRandomString] = useState<string>('');

  useEffect(() => {
    refresh();
  }, []);

  const refresh = (options: GeneratorOptions = initialOptions) => {
    const varchar = generator(options);
    setRandomString(varchar);
  };

  const onValuesChange = (_: any, allValues: GeneratorOptions) => {
    refresh(allValues);
  };

  const onFinish = (values: GeneratorOptions) => {
    refresh(values);
  };

  return (
    <PageContainer title={false} className="mx-auto max-w-screen-lg px-4 pt-4">
      <Form layout="horizontal" initialValues={initialOptions} onValuesChange={onValuesChange} onFinish={onFinish}>
        <Card actions={[<Button key="1" htmlType="submit" icon={<Iconify icon="material-symbols:refresh" />} />]}>
          <div className="grid grid-cols-1 sm:grid-cols-12">
            <Form.Item name="numeric" valuePropName="checked" className="sm:col-span-6">
              <Checkbox>数字 0 ～ 9</Checkbox>
            </Form.Item>
            <Form.Item name="lowercase" valuePropName="checked" className="sm:col-span-6">
              <Checkbox>小写字母 a ~ z</Checkbox>
            </Form.Item>
            <Form.Item name="uppercase" valuePropName="checked" className="sm:col-span-6">
              <Checkbox>大写字母 A ~ Z</Checkbox>
            </Form.Item>
            <div className="flex sm:col-span-6">
              <Form.Item name="symbol" valuePropName="checked" className="basis-20">
                <Checkbox>符号</Checkbox>
              </Form.Item>
              <Form.Item name="symbolChars" className="flex-1">
                <Input />
              </Form.Item>
            </div>
          </div>
          <div>
            <Form.Item name="length" label="长度">
              <Slider min={6} max={64} step={1} marks={{ 6: '6', 8: '8', 16: '16', 32: '32', 64: '64' }} />
            </Form.Item>
          </div>
          <div>
            <InputCopyable value={randomString} />
          </div>
        </Card>
      </Form>
    </PageContainer>
  );
};
