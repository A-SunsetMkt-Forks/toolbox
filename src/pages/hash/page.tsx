import InputCopyable from '@/components/InputCopyable.tsx';
import { PageContainer } from '@ant-design/pro-components';
import { Card, Input } from 'antd';
import { MD5, SHA1, SHA256, SHA512 } from 'crypto-js';
import type { FC } from 'react';
import { useEffect, useState } from 'react';

interface HashResult {
  md5?: string;
  sha1?: string;
  sha256?: string;
  sha512?: string;
}

export const Component: FC = () => {
  const defaultHashResult: HashResult = {
    md5: '',
    sha1: '',
    sha256: '',
    sha512: '',
  };

  const [plainText, setPlainText] = useState<string>('');
  const [hashResult, setHashResult] = useState<HashResult | null>(defaultHashResult);
  useEffect(() => {
    if (!plainText || plainText === '') {
      setHashResult((defaultHashResult) => defaultHashResult);
      return;
    }
    const md5 = MD5(plainText).toString();
    const sha1 = SHA1(plainText).toString();
    const sha256 = SHA256(plainText).toString();
    const sha512 = SHA512(plainText).toString();
    setHashResult((hashResult) => {
      return { ...hashResult, md5, sha1, sha256, sha512 };
    });
  }, [plainText]);
  return (
    <PageContainer title={false} className="mx-auto max-w-screen-md pt-4">
      <Card>
        <div className="mb-8">
          <Input.TextArea
            value={plainText}
            rows={2}
            placeholder="原文"
            onChange={(event: any) => {
              setPlainText(event.target.value);
            }}
            showCount
          />
        </div>

        <InputCopyable tag="MD5" value={hashResult?.md5} />
        <InputCopyable tag="SHA1" value={hashResult?.sha1} />
        <InputCopyable tag="SHA256" value={hashResult?.sha256} />
        <InputCopyable tag="SHA512" value={hashResult?.sha512} />
      </Card>
    </PageContainer>
  );
};
