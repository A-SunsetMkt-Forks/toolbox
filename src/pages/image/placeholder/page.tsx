import {
  FormControlRender,
  PageContainer,
  pickControlPropsWithId,
  ProForm,
  ProFormDigit,
  ProFormSelect,
  ProFormText,
} from '@ant-design/pro-components';
import type { TabsProps } from 'antd';
import { Card, ColorPicker, Form, Tabs } from 'antd';
import type { FC } from 'react';
import { useEffect, useState } from 'react';

import type { PlaceholderConfig } from '@/components/placeholder';
import DummyImage from '@/components/placeholder/DummyImage.tsx';
import HrefLuImage from '@/components/placeholder/HrefLuImage.tsx';
import LocalImage from '@/components/placeholder/LocalImage.tsx';
import PlaceholderImage from '@/components/placeholder/PlaceholderImage.tsx';

const defaultValues = {
  width: 320,
  height: 320,
  suffix: '.png',
  bgColor: '#8c8c8c',
  textColor: '#f0f0f0',
  text: 'img',
};

const suffixOptions = [
  { label: '.png', value: '.png' },
  { label: '.jpg', value: '.jpg' },
  { label: '.gif', value: '.gif' },
];

export const Component: FC = () => {
  const [form] = Form.useForm();
  const [placeholderConfig, setPlaceholderConfig] = useState<PlaceholderConfig>(defaultValues);

  const bgColor = Form.useWatch('bgColor', { form, preserve: true });
  const textColor = Form.useWatch('textColor', { form, preserve: true });

  useEffect(() => {
    const values = form.getFieldsValue();
    setPlaceholderConfig(values);
  }, [bgColor, form, textColor]);

  const tabsProps: TabsProps = {
    defaultActiveKey: 'LocalImage',
    items: [
      {
        key: 'LocalImage',
        label: 'Local',
        children: <LocalImage {...placeholderConfig} />,
      },
      {
        key: 'HrefLuImage',
        label: 'iph.href.lu',
        children: <HrefLuImage {...placeholderConfig} />,
      },
      {
        key: 'PlaceholderImage',
        label: 'via.placeholder.com',
        children: <PlaceholderImage {...placeholderConfig} />,
      },
      {
        key: 'DummyImage',
        label: 'dummyimage.com',
        children: <DummyImage {...placeholderConfig} />,
      },
    ],
  };
  return (
    <PageContainer title={false} className="mx-auto max-w-screen-lg p-4">
      <Card>
        <ProForm
          submitter={false}
          form={form}
          initialValues={defaultValues}
          onValuesChange={(_, values) => {
            setPlaceholderConfig(values);
          }}
        >
          <div className="grid grid-cols-2 gap-2 sm:grid-cols-3">
            <ProFormDigit label="宽度" name="width" fieldProps={{ size: 'large' }} />
            <ProFormDigit label="高度" name="height" fieldProps={{ size: 'large' }} />
            <ProFormSelect
              label="后缀名"
              name="suffix"
              options={suffixOptions}
              allowClear={false}
              fieldProps={{ size: 'large' }}
            />
            <ProForm.Item label="背景颜色" name="bgColor">
              <FormControlRender>
                {(itemProps) => {
                  return (
                    <ColorPicker
                      size="large"
                      showText
                      disabledAlpha
                      className="w-full justify-start"
                      {...pickControlPropsWithId(itemProps)}
                      onChange={(color) => {
                        form.setFieldValue('bgColor', color.toHexString());
                      }}
                    />
                  );
                }}
              </FormControlRender>
            </ProForm.Item>
            <ProForm.Item label="文本颜色" name="textColor">
              <FormControlRender>
                {(itemProps) => {
                  return (
                    <ColorPicker
                      size="large"
                      showText
                      disabledAlpha
                      className="w-full justify-start"
                      {...pickControlPropsWithId(itemProps)}
                      onChange={(color) => {
                        form.setFieldValue('textColor', color.toHexString());
                      }}
                    />
                  );
                }}
              </FormControlRender>
            </ProForm.Item>
            <ProFormText label="文本" name="text" allowClear={false} fieldProps={{ size: 'large' }} />
          </div>
        </ProForm>
        <div>
          <Tabs {...tabsProps} />
        </div>
      </Card>
    </PageContainer>
  );
};
