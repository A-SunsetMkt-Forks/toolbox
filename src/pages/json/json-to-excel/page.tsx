import JSON5 from 'json5';
import _ from 'lodash';
import { useEffect, useState } from 'react';
import type { IAceOptions } from 'react-ace';
import AceEditor from 'react-ace';
import ReactDataSheet from 'react-datasheet';

import 'ace-builds/src-noconflict/ext-searchbox';
import 'ace-builds/src-noconflict/mode-json5';
import 'ace-builds/src-noconflict/mode-yaml';
import 'ace-builds/src-noconflict/theme-textmate';

import 'react-datasheet/lib/react-datasheet.css';

export interface GridElement extends ReactDataSheet.Cell<GridElement, number | string | null> {
  value: number | string | null;
}

export const Component = () => {
  const [json, setJson] = useState<string>('{}');
  const [data, setData] = useState<GridElement[][]>([]);
  const [jsonOptions] = useState<IAceOptions>({
    useWorker: false,
    showLineNumbers: true,
    tabSize: 2,
    wrap: true,
    fontSize: 14,
  });
  useEffect(() => {
    try {
      const result = JSON5.parse(json);
      console.log(result);
      if (Array.isArray(result)) {
        let columns: any[] = [];
        result.forEach((item) => {
          const keys = Object.keys(item);
          keys.forEach((key) => {
            columns.push({ value: key, readOnly: true });
          });
        });
        columns = _.uniqWith(columns, (a, b) => {
          return a.value === b.value;
        });

        const rows: any[] = [];
        result.forEach((item) => {
          const row = columns.map((column) => {
            const value = item[column.value] || null;
            return { value, readOnly: true };
          });
          rows.push(row);
        });
        // @ts-ignore
        setData([columns, ...rows]);
      }
    } catch (e) {
      if (e instanceof SyntaxError) {
        console.error(e.message);
      }
    }
  }, [json]);

  return (
    <div>
      <div className="h-[300px] w-full">
        <AceEditor
          theme="textmate"
          placeholder="JSON"
          setOptions={jsonOptions}
          value={json}
          onChange={setJson}
          mode="json5"
          width="100%"
          height="100%"
        />
      </div>
      <div className="mt-6 w-full">
        <ReactDataSheet data={data} valueRenderer={(cel) => cel.value} />
      </div>
    </div>
  );
};
