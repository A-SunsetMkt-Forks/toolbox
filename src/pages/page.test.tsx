import { render, screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";

import { MemoryRouter } from "react-router-dom";
import { Component } from "./page.tsx";

describe("Input", async () => {
  it("app", () => {
    render(
      <MemoryRouter>
        <Component />
      </MemoryRouter>,
    );

    const element = screen.getByRole("link", { name: "JSON" });
    expect(element).toBeInTheDocument();
  });
});
