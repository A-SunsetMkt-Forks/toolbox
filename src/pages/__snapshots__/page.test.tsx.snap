// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders learn react link 1`] = `
<a
  href="http://antfu.me"
>
  Anthony Fu
</a>
`;
